relay-general/src/store/normalize/measurements.rs

```diff
@@ -0,0 +1,357 @@
+//! Contains the measurements normalization code
+//!
+//! This module is responsible for ensuring the measurements interface is only present for transaction
+//! events, and generating operation breakdown measurements.
+//!
+
+use std::collections::HashMap;
+
+use crate::protocol::{Event, EventType, Measurement, Measurements, Timestamp};
+use crate::types::Annotated;
+
+#[derive(Clone, Debug)]
+struct TimeWindowSpan {
+    start_timestamp: Timestamp,
+    end_timestamp: Timestamp,
+}
+
+impl TimeWindowSpan {
+    fn new(start_timestamp: Timestamp, end_timestamp: Timestamp) -> Self {
+        if end_timestamp < start_timestamp {
+            return TimeWindowSpan {
+                start_timestamp: end_timestamp,
+                end_timestamp: start_timestamp,
+            };
+        }
+
+        TimeWindowSpan {
+            start_timestamp,
+            end_timestamp,
+        }
+    }
+}
+
+type OperationName = String;
+
+#[derive(PartialEq, Eq, Hash)]
+enum OperationBreakdown {
+    Emit(OperationName),
+    DoNotEmit(OperationName),
+}
+
+type OperationNameIntervals = HashMap<OperationBreakdown, Vec<TimeWindowSpan>>;
+
+fn merge_intervals(mut intervals: Vec<TimeWindowSpan>) -> Vec<TimeWindowSpan> {
+    // sort by start_timestamp in ascending order
+    intervals.sort_unstable_by(|a, b| a.start_timestamp.partial_cmp(&b.start_timestamp).unwrap());
+
+    intervals.into_iter().fold(
+        vec![],
+        |mut merged, current_interval| -> Vec<TimeWindowSpan> {
+            // merged is a vector of disjoint intervals
+
+            if merged.is_empty() {
+                merged.push(current_interval);
+                return merged;
+            }
+
+            let mut last_interval = merged.last_mut().unwrap();
+
+            if last_interval.end_timestamp < current_interval.start_timestamp {
+                // if current_interval does not overlap with last_interval,
+                // then add current_interval
+                merged.push(current_interval);
+                return merged;
+            }
+
+            // current_interval and last_interval overlaps; so we merge these intervals
+
+            // invariant: last_interval.start_timestamp <= current_interval.start_timestamp
+
+            last_interval.end_timestamp =
+                std::cmp::max(last_interval.end_timestamp, current_interval.end_timestamp);
+
+            merged
+        },
+    )
+}
+
+/// Ensure measurements interface is only present for transaction events, and emit operation breakdown measurements
+pub fn normalize_measurements(event: &mut Event, operation_name_breakdown: &Option<Vec<String>>) {
```

Comment:
A more idiomatic signature (also unwrapping the option) would be:

```suggestion
pub fn normalize_measurements(event: &mut Event, operation_name_breakdown: &[String]) {
```